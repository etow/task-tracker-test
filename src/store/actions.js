import taskRepository from '../repositories/taskRepository';
import taskCategoryRepository from '../repositories/taskCategoryRepository';
import {
  SET_TASKS_CATEGORIES,
  SET_TASKS_CATEGORIES_DATA,
  SET_TASKS,
  ADD_TASK,
  DELETE_TASK,
  SET_TASK_TO_EDIT,
  SET_TASK,
  FETCH_TASKS_CATEGORIES,
  FETCH_TASKS,
  CREATE_TASK,
  UPDATE_TASK,
  ROLLBACK_STATE,
  BACKUP_STATE,
  UPDATE_TASKS,
} from './constants';

export default {
  [FETCH_TASKS_CATEGORIES]: ({ commit }) => {
    return taskCategoryRepository.get()
      .then((response) => {
        commit(SET_TASKS_CATEGORIES, response.data);
        commit(SET_TASKS_CATEGORIES_DATA, response.data);
      });
  },
  [FETCH_TASKS]: ({ commit }) => {
    return taskRepository.get()
      .then((response) => commit(SET_TASKS, response.data));
  },
  [CREATE_TASK]: ({ commit, state }, newTask) => {
    commit(BACKUP_STATE, state);
    const task = {
      ...newTask,
      id: Math.random().toString(36).substr(2, 6),
      order: state.tasks[newTask.category].length,
    };
    commit(ADD_TASK, task);
    return taskRepository.create(task);
  },
  [DELETE_TASK]: ({ commit, state }, {task, category}) => {
    commit(BACKUP_STATE, state);
    commit(DELETE_TASK, {task, category});
    commit(SET_TASK_TO_EDIT, {});
    return taskRepository.delete(task.id);
  },
  [UPDATE_TASK]: ({ commit, state }, {task, prevCategory}) => {
    commit(BACKUP_STATE, state);
    if (prevCategory && task.category !== prevCategory) {
      commit(DELETE_TASK, { task, category: prevCategory });
    }
    const targetIndex = state.tasks[task.category].findIndex((aTask) => aTask.id === task.id);
    if (targetIndex >= 0) {
      commit(SET_TASK, { task, targetIndex });
    } else {
      const order = state.tasks[task.category].length;
      commit(ADD_TASK, { ...task, order });
    }
    commit(SET_TASK_TO_EDIT, {})
    return taskRepository.update(task);
  },
  [UPDATE_TASKS]: ({ commit, state }, { tasks, category }) => {
    commit(BACKUP_STATE, state);
    tasks.forEach((task, index) => {
      task.category = category;
      task.order = index;
    });
    commit(UPDATE_TASKS, { tasks, category });

    return taskRepository.update(tasks);
  },
  [SET_TASK_TO_EDIT]: ({ commit, state}, task) => {
    commit(BACKUP_STATE, state);
    commit(SET_TASK_TO_EDIT, task);
  },
  [ROLLBACK_STATE]: ({ commit }) => {
    commit(ROLLBACK_STATE);
  },
};
