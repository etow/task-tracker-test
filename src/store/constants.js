

/* mutations and actions */
export const ADD_TASK = 'ADD_TASK';
export const SET_TASKS = 'SET_TASKS';
export const SET_TASKS_CATEGORIES = 'SET_TASKS_CATEGORIES';
export const SET_TASKS_CATEGORIES_DATA = 'SET_TASKS_CATEGORIES_DATA';
export const SET_TASK = 'SET_TASK';
export const SET_TASK_TO_EDIT = 'SET_TASK_TO_EDIT';
export const DELETE_TASK = 'DELETE_TASK';
export const FETCH_TASKS_CATEGORIES = 'FETCH_TASKS_CATEGORIES';
export const FETCH_TASKS = 'FETCH_TASKS';
export const CREATE_TASK = 'CREATE_TASK';
export const UPDATE_TASK = 'UPDATE_TASK';
export const ROLLBACK_STATE = 'ROLLBACK_STATE';
export const BACKUP_STATE = 'BACKUP_STATE';
export const UPDATE_TASKS = 'UPDATE_TASKS';
export const UPDATE_TASK_ACTIVITY = 'UPDATE_TASK_ACTIVITY';
