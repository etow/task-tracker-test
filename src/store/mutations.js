import {
  ADD_TASK,
  SET_TASKS,
  SET_TASKS_CATEGORIES,
  SET_TASKS_CATEGORIES_DATA,
  SET_TASK,
  SET_TASK_TO_EDIT,
  DELETE_TASK,
  ROLLBACK_STATE,
  BACKUP_STATE,
  UPDATE_TASKS,
  UPDATE_TASK_ACTIVITY
} from './constants';

export default {
  [ADD_TASK]: (state, task) => {
    state.tasks[task.category].push(task);
  },
  [SET_TASKS]: (state, tasks) => {
    Object.keys(state.tasks)
      .forEach((category) => {
        const tasksByCategory = tasks.filter((task) => task.category === category);
        state.tasks[category] = tasksByCategory.sort((a, b) => a.order - b.order);
      });
  },
  [SET_TASKS_CATEGORIES]: (state, categories) => {
    categories.forEach((cat) => state.tasks[cat.name] = []);
  },
  [SET_TASKS_CATEGORIES_DATA]: (state, categories) => {
    categories.forEach((cat) => state.categories[cat.name] = cat);
  },
  [SET_TASK]: (state, { task, targetIndex }) => {
    const index = targetIndex || state.tasks[task.category].findIndex((aTask) => aTask.id === task.id);
    state.tasks[task.category][index] = task;
  },
  [SET_TASK_TO_EDIT]: (state, task) => {
    state.taskToEdit = task;
  },
  [DELETE_TASK]: (state, {task, category}) => {
    state.tasks[category] = state.tasks[category].filter((aTask) => aTask.id !== task.id);
  },
  [ROLLBACK_STATE]: (state) => {
    Object.assign(state, JSON.parse(JSON.stringify(state.prevState)));
  },
  [BACKUP_STATE]: (state, currentState) => {
    state.prevState = JSON.parse(JSON.stringify(currentState));
  },
  [UPDATE_TASKS]: (state, {tasks, category}) => {
    state.tasks[category] = tasks;
  },
  [UPDATE_TASK_ACTIVITY]: (state, tasks) => {
    if (!Array.isArray(tasks)) {
      tasks = [tasks];
    }
    tasks.forEach((task) => {
      const found = state.tasks[task.category].find((aTask) => aTask.id === task.id);
      found.activity = task.activity;
    });
  }
};
