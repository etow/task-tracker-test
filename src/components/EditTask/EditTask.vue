<template>
  <div>
    <ElDialog
      :model-value="showDialog"
      :before-close="handleClose"
      :show-close="false"
      width="30%"
      custom-class="form-dialog"
      center>
      <ElTabs v-model="activeTab">
        <ElTabPane label="Edit task" name="edit">
          <EditTaskForm
            @save="handleClose"
            @cancel="handleClose"
            @delete="handleClose" />
        </ElTabPane>
        <ElTabPane label="Activity" name="activity">
          <Activity />
        </ElTabPane>
      </ElTabs>
    </ElDialog>
  </div>
</template>

<script>
import {
  ElDialog,
  ElTabs,
  ElTabPane,
} from 'element-plus';
import Activity from '../Activity/Activity.vue';
import EditTaskForm from '../EditTaskForm/EditTaskForm.vue';
import { SET_TASK_TO_EDIT } from '../../store/constants';

export default {
  data() {
    return {
      activeTab: 'edit',
      showDialog: false,
    };
  },
  components: {
    ElDialog,
    ElTabs,
    ElTabPane,
    Activity,
    EditTaskForm,
  },
  watch: {
    taskToEdit: {
      handler(task) {
        if (Object.keys(task).length) {
          this.showDialog = true;
          this.activeTab = 'edit';
        }
      },
      inmediate: true,
    },
  },
  computed: {
    taskToEdit() {
      return this.$store.state.taskToEdit;
    },
  },
  methods: {
    handleClose() {
      this.showDialog = false;
      this.$nextTick(() => {
        this.$store.dispatch(SET_TASK_TO_EDIT, {});
      });
    },
  },
};
</script>
<style lang="scss" scoped>
  ::v-deep(.el-dialog__header) {
    display: none;
  }
  ::v-deep(.form-dialog) {
    border-radius: 20px;
  }
  ::v-deep(.el-tabs__content) {
    padding-top: 20px;
  }
</style>
