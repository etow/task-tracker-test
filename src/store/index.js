import { createStore } from 'vuex';

import mutations from './mutations';
import actions from './actions';

export default createStore({
  state: {
    tasks: {},
    categories: {},
    taskToEdit: {},
    prevState: {},
  },
  mutations,
  actions,
});
