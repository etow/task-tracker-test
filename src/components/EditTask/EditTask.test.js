import { shallowMount, config } from '@vue/test-utils';
import { createStore } from 'vuex';
import EditTask from './EditTask.vue';

import { SET_TASK_TO_EDIT } from '../../store/constants';

jest.mock('element-plus', () => ({
  ElMessage: jest.fn(),
}));

const tasks = [
  { name: 'test1', category: 'Planned', order: 0, estimate: 0, activity: {} },
  { name: 'test2', category: 'Planned', order: 1, estimate: 0, activity: {} },
  { name: 'test3', category: 'Planned', order: 2, estimate: 0, activity: {} },
];

beforeAll(() => {
  config.renderStubDefaultSlot = true;
});

afterAll(() => {
  config.renderStubDefaultSlot = false;
});

describe('EditTask.vue', () => {
  it('renders correctly', async () => {
    // mock store
    const [store] = mockStore();
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('closes dialog', async () => {
    const [store] = mockStore();
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });
    store.state.taskToEdit = tasks[0];
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.showDialog).toBe(true);
    wrapper.vm.handleClose();
    expect(wrapper.vm.showDialog).toBe(false);
  });
});


const createWrapper = (decorator = (n) => n) => {
  const dataToMount = {
    global: {
      stubs: {
        ElDialog: true,
        ElTabs: true,
        ElTabPane: true,
        EditTaskForm: true,
        Activity: true,
      },
    },
  };

  decorator(dataToMount);
  return shallowMount(EditTask, dataToMount);
};

function mockStore({ set_task_to_edit, taskToEdit } = {}) {
  const state = {
    tasks: tasks,
    categories: {
      Planned: { name: 'Planned', color: 'red' },
      'In Progress': { name: 'In Progress', color: 'red' },
      Completed: { name: 'Completed', color: 'red' },
    },
    taskToEdit: taskToEdit || {},
  };

  const actions = {
    [SET_TASK_TO_EDIT]: set_task_to_edit || jest.fn(() => Promise.resolve()),
  }

  const store = createStore({
    state,
    actions,
  });

  return [store, { actions }];
}
