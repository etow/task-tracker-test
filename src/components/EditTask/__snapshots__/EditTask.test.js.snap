// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditTask.vue renders correctly 1`] = `
<div>
  <el-dialog-stub model-value="false" before-close="function () { [native code] }" show-close="false" width="30%" custom-class="form-dialog" center="">
    <el-tabs-stub modelvalue="edit">
      <el-tab-pane-stub label="Edit task" name="edit">
        <edit-task-form-stub></edit-task-form-stub>
      </el-tab-pane-stub>
      <el-tab-pane-stub label="Activity" name="activity">
        <activity-stub></activity-stub>
      </el-tab-pane-stub>
    </el-tabs-stub>
  </el-dialog-stub>
</div>
`;
